clippy_lints/src/large_future.rs

```diff
@@ -0,0 +1,78 @@
+use clippy_utils::{diagnostics::span_lint, ty::implements_trait};
+use rustc_hir::{Expr, ExprKind, LangItem, MatchSource, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_target::abi::Size;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// It checks for the size of a `Future` created by `async fn` or `async {}`.
+    ///
+    /// ### Why is this bad?
+    /// Due to the current [unideal implemention](https://github.com/rust-lang/rust/issues/69826) of `Generator`,
+    /// large size of a `Future` may cause stack overflows.
+    ///
+    /// ### Example
+    /// ```rust
+    /// async fn wait(f: impl std::future::Future<Output = ()>) {}
+    ///
+    /// async fn big_fut(arg: [u8; 1024]) {}
+    ///
+    /// pub async fn test() {
+    ///     let fut = big_fut([0u8; 1024]);
+    ///     wait(fut).await;
+    /// }
+    /// ```
+    ///
+    /// `Box::pin` the big future instead.
+    ///
+    /// ```txt
+    /// let fut = Box::pin(big_fut([0u8; 1024]));
+    /// ```
+    #[clippy::version = "1.68.0"]
+    pub LARGE_FUTURE,
```

Comment:
Small naming nit: Lint names should [ideally be plural](https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints). So `LARGE_FUTURES` would be a better name. You can try renaming the lint with `cargo dev`. I don't remember the exact comment, but you should find it in the help menu. It'll add some tests, for renamed lints, which will need to be removed, but it does the whole file renaming thing for you :)